06_gpu_and_ml/text-to-video/mochi.py

```diff
@@ -28,338 +27,147 @@
+    @modal.build()
+    @modal.enter()
+    def load_model(self):
+        device = "cuda" if torch.cuda.is_available() else "cpu"
+        self.pipe = MochiPipeline.from_pretrained(
+            "genmo/mochi-1-preview", 
+            torch_dtype=torch.bfloat16, 
+        )
+        self.pipe.enable_model_cpu_offload()
+        self.pipe.enable_vae_tiling()
 
+    @modal.method()
+    def generate(self, 
+        prompt, 
+        negative_prompt = "",
+        num_inference_steps = 200,
+        guidance_scale = 4.5,
+        num_frames = 19,
+    ):
+        frames = self.pipe(
+            prompt=prompt, 
+            negative_prompt=negative_prompt,
+            num_inference_steps=num_inference_steps, 
+            guidance_scale=guidance_scale,
+            num_frames=num_frames,
+        ).frames[0]
+        
+        # save to disk using prompt as filename
+        for char in string.punctuation:
+            prompt = prompt.replace(char, "")
+        prompt = prompt.replace(" ", "_")
+        prompt = prompt[:230] # since filenames can't be longer than 255 characters
+        mp4_name = str(int(time.time())) + "_" + prompt + ".mp4"
+        
+        export_to_video(frames, Path(OUTPUTS_PATH) /  mp4_name)
+        outputs.commit()
+        return mp4_name
 
 # ## Running Mochi inference
 
 # We can trigger Mochi inference from our local machine by running the code in
 # the local entrypoint below.
 
-# It ensures the model is downloaded to a remote volume,
-# spins up a new replica to generate a video, also saved remotely,
-# and then downloads the video to the local machine.
+# It ensures the model is downloaded into the image,
+# spins up a replica to generate a video,
+# and then downloads that video to the local machine.
 
 # You can trigger it with:
 
 # ```bash
 # modal run --detach mochi
 # ```
-
+# 
+# Optional command line flags are:
+# --prompt="your prompt"
+# --negative-prompt="your negative prompt"
+# --num-inference-steps=200
+# --guidance-scale=4.5
+# --num-frames=19
+# 
+# Such as:
+# ```bash
+# modal run --detach mochi --prompt="a cat playing drums in a jazz ensemble" --num-inference-steps=64
+# ```
 
 @app.local_entrypoint()
 def main(
-    prompt: str = "A cat playing drums in a jazz ensemble",
-    num_inference_steps: int = 200,
-):
-    from pathlib import Path
-
+    prompt = "Close-up of a chameleon's eye, with its scaly skin changing color. Ultra high resolution 4k.",
+    negative_prompt = "",
+    num_inference_steps = 200,
+    guidance_scale = 4.5,
+    num_frames = 19, # produces ~1s of video
+    ): 
+    
     mochi = Mochi()
+    mp4_name = mochi.generate.remote(
+        prompt=str(prompt), 
+        negative_prompt=str(negative_prompt),
+        num_inference_steps=int(num_inference_steps), 
+        guidance_scale=float(guidance_scale),
+        num_frames=int(num_frames),
+    )
+    print("🍡 video saved to volume at "+mp4_name)
```

Comment:
cleaner:

```suggestion
    print("🍡 video saved to volume at", mp4_name)
```

I tend to prefer f-strings, since they are equally, if not more, readable and are more extensible.